### PR TITLE
Method to fetch all intermines using registry

### PR DESCRIPTION
**Pull request by LakshSingla**
### Please describe your PR:
Created Registry class in order to provide helper methods for the Intermine-Registry and acts as an abstraction between client and the registry. Refers #22  
### Review checklist:

 Before merging, confirm the following tasks have been executed

- [x] Relevant docs have been added/updated

If this is for a new release, make sure the following steps have occurred:

- [ ] Update [CHANGES](CHANGES) to reflect the new updates.
- [ ] Bump version in:
    - [ ] bower.json (do this manually by editing bower.json)
    - [ ] package.json - do this by running `npm version newversion` - e.g. `npm version 3.17.0`. This will automatically tag and commit the version bump.

This is not an exhaustive list - if you spot something else strange please bring it up!

See [release-procedure](release-procedure.md) for info on releasing an update.

### CODE REVIEW

**Comment by LakshSingla** [comment]
@yochannah I have added some docs and refactored borrowing style from the `Service` class. The `makeRequest` method in `Registry` class will get updated as and when different use cases for the class will occur :smile:

File: src/registry.coffee [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+  # Default error handler function to provide error stream if present
+  # or fallback to stdout
+  # @param [Any] e Logs the parameter provided under error stream if present
+  #   or under the logging stream
+  errorHandler: (e) ->
+    f = console.error ? console.log
+    f e
+
+ # Concatenates relative path with root and returns final path to call
```
Comment (yochannah) [inline src/registry.coffee]: Tiny note - this line was indented differently from the rest of the block and it made the docs build fail 😆 not a big thing though

File: src/registry.coffee [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      dataType: dataType
+      success: cb
+      error: errBack
+      type: method
+      url: @makePath [INSTANCES_PATH]
+
+    withCB cb, http.doReq.call this, opts
+
+  # Fetches instances of all known registry information
+  # @param[->] cb A function to be attatched to the returned promise
```
Comment (yochannah) [inline src/registry.coffee]: another docs-related note - the space is needed between @param and [->] for it to pick up correctly. (who thought taking js and making it whitespace sensitive was a good idea anyway? :laughing:

**Review by yochannah — COMMENTED** [review]
@LakshSingla I have added a couple of minor comments, but they are *very* minor :) 

I've made a few additions and changes to master that includes documentation for how to generate the docs, and allows us to build imjs without running the test (which require an intermine to be set up locally). Maybe you want to pull those changes in as well? 

re:testing this code: Could you possible share the code you used to test the registry methods, and instructions on how to run it?

**Comment by LakshSingla** [comment]
@yochannah I used CoffeeScript REPL to test the code (which I think for sure is an unsustainable method). I was reading up on Travis (as mentioned by you in the issue), to find a way to running the tests. 
I will make the required changes and bring the branch upto master at the earliest. :)

**Comment by LakshSingla** [comment]
@yochannah I have brought my fork upto date with the origin's(this repository) fork. I am currently looking up on the changes, and will get back to you. 
Thanks for pointing these issues out, I will be more careful about them next time.
**UPDATE**: I was able to set up Travis successfully :tada:, will add unit tests for the `Registry` class by tomorrow. Is there any naming or numbering convention to be followed :thinking: , some of the tests are prepended by a number while others are not.

**Comment by LakshSingla** [comment]
@yochannah I was reviewing the API once again and I missed some query parameters that can be passed down. I am adding support for them, and will write the tests for all of them together. 
**UPDATE**: Added support for query paramters, will add test cases by tonight

**Comment by yochannah** [comment]
@LakshSingla This all sounds great! 

Don't worry about numbering! I think it might be automatic?

**Comment by LakshSingla** [comment]
@yochannah I am having some difficulty in adding new test cases. I pushed my commit to the branch `dev` having my test case in `test/mocha/registry.coffee`. On checking the Travis console, it is not detecting my test. Is there any other location where I need to make a change, in order for it to be detected.

**Comment by yochannah** [comment]
hmmmm, this should be interesting to hunt down :laughing: 

it looks like the gruntfile should detect all tests in the folder by default - 

```json
      tests: {
        src: 'test/mocha/*.coffee',
        dest: 'test/compiled'
},```

so it's not that. I'll continue investigating.

**Comment by LakshSingla** [comment]
@yochannah If it helps [Travis build](https://travis-ci.com/LakshSingla/imjs)

**Comment by LakshSingla** [comment]
@yochannah I am unable to get  asynchronous tests to pass, even though I am explicitly rejecting promise. Apart from that, the tests for helper functions are written and complete.
One more thing, I have renamed the function from `fetchAllMines` to `fetchMines` because one can fetch specific mines passing in custom paramters. 
I will update the README to reflect my findings during this process and open another PR by the evening. I will also dig into how other asynchronous tests are written.

**Comment by yochannah** [comment]
@LakshSingla This is great!!! :100: 

Here's the code I used to verify things worked as expected: 

```html
<!DOCTYPE html>
<html lang="en" dir="ltr">
  <head>
    <meta charset="utf-8">
    <title></title>
  </head>
  <body>
    <script src="im.js"></script>
    <script>
    var registry = new imjs.Registry();

    //Happy paths

    // fetchmines default / prod
    registry.fetchMines().then(function(response) {
      console.log("all prod mines (default)",response.instances);
    });

    // fetchmines specify prod
    registry.fetchMines([],["prod"]).then(function(response) {
      console.log("all prod mines (with arg)",response.instances);
    });

    //fetchmines with search
    registry.fetchMines(["D. melanogaster"]).then(function(response) {
      console.log("mines with flies in their descriptions",response.instances);
    });

    //fetchmines with multiple search params
    registry.fetchMines(["H. sapiens", "legume"]).then(function(response) {
      console.log("mines with humans or beans in their descriptions",response.instances);
    });

    //fetchmines with search and dev enabled
    registry.fetchMines(["yeast"], ["dev"]).then(function(response) {
      console.log("all yeast dev mines",response.instances);
    });

    //fetchmines all dev no prod
    registry.fetchMines([], ["dev"]).then(function(response) {
      console.log("all dev mines",response.instances);
    });

    //fetchmines all - prod and dev
    registry.fetchMines([], ["all"]).then(function(response) {
      console.log("all mines",response.instances);
    });

    var mycallback = function() {
      console.log("The callback worked if this prints!");
    }

    //fetchmines prod + callback
    registry.fetchMines([], [], mycallback).then(function(response) {
      console.log("all mines",response.instances);
    });

/////////////////////////////////////
//sad paths

    //non-existent search term
    registry.fetchMines(["Kittens!"]).then(function(response) {
      console.log("non-existent-search-term",response.instances);
    });

    //not dev, not prod, not all...
    registry.fetchMines([],["bad-arg"]).then(function(response) {
      console.log("non-existent-search-term",response.instances);
    });

    </script>
  </body>
</html>

```

image

All of the checks I wrote passed perfectly and I was extra impressed to see the nice error handling for dev/prod mines. Nicely done!

**Pull request merged**